ingestor-default/src/test/scala/za/co/absa/hyperdrive/ingestor/implementation/utils/TestKafkaUtilDockerTest.scala

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.ingestor.implementation.utils
+
+import java.time.Duration
+import java.util
+import java.util.UUID.randomUUID
+import java.util.{Collections, Properties}
+
+import org.apache.kafka.clients.admin.{AdminClient, AdminClientConfig, NewTopic}
+import org.apache.kafka.clients.consumer.{ConsumerRecord, KafkaConsumer}
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.serialization.{StringDeserializer, StringSerializer}
+import org.scalatest.{AppendedClues, BeforeAndAfter, FlatSpec, Matchers}
+import org.testcontainers.containers.KafkaContainer
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+class TestKafkaUtilDockerTest extends FlatSpec with Matchers with BeforeAndAfter with AppendedClues {
+
+  private val confluentPlatformVersion = "5.3.4" // should be same as kafka.avro.serializer.version property in pom file
+  private val kafka = new KafkaContainer(confluentPlatformVersion)
+  private val kafkaSufficientTimeout = Duration.ofSeconds(5L)
+  private val kafkaInsufficientTimeout = Duration.ofMillis(1L)
+  private val topic = "test-topic"
+
+  before{
+    kafka.start()
+  }
+
+  after {
+    kafka.stop()
+  }
+
+  "getAllAvailableMessages" should "get all available messages" in {
+    // given
+    val partitions = 3
+    createTopic(kafka, topic, partitions)
+    val producer = createProducer(kafka)
+    val messages = (1 to 100).map(i => s"message_${i}")
+    produceData(producer, messages, topic, partitions)
+
+    val consumer = createConsumer(kafka)
+    val topicPartitions = KafkaUtil.getTopicPartitions(consumer, topic)
+    val offsets = consumer.endOffsets(topicPartitions.asJava).asScala.toMap.mapValues(_.asInstanceOf[Long])
+
+    // when
+    implicit val kafkaConsumerTimeout: Duration = kafkaSufficientTimeout
+    val records = KafkaUtil.getMessagesAtLeastToOffset(consumer, offsets)
+
+    // then
+    val actualMessages = records.map(_.value()).toList.sorted
+    actualMessages should contain theSameElementsAs messages
+  }
+
+  it should "get all available messages, even if polling is required multiple times" in {
```

Comment:
How do you know that polling is required multiple times ?

Reply:
If I remove the while loop in the implementation, this test should fail. I don't know exactly why it fails, but it has to do with "segment.ms" -> "100". This instructs kafka to create a new segment (that's a building block of a partition) every 100ms. And somehow that causes the consumer not to get all available messages in one poll, but it needs multiple polls.

The first test in this suite "get all available messages" doesn't fail even if I remove the while loop.

Reply:
Interesting, I hoped that you will get to the "segment.ms". 
Btw you can use params that you know what they are doing like props.put("max.poll.records", "10")